### ISSUE TITLE
t.w.static.addSlash does not work on Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8169]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8169|
|Type|defect|
|Created|2016-01-09 04:29:41Z|
|Branch|https://github.com/twisted/twisted/tree/addSlash-8169-3|

addSlash, called when you access a folder without a trailing slash when browsing a ```static.File```, doesn't work on Python 3.

```
Traceback (most recent call last):
  File "/home/hawkowl/venvs/py3/lib/python3.4/site-packages/twisted/protocols/basic.py", line 571, in dataReceived
    why = self.lineReceived(line)
  File "/home/hawkowl/venvs/py3/lib/python3.4/site-packages/twisted/web/http.py", line 1720, in lineReceived
    self.allContentReceived()
  File "/home/hawkowl/venvs/py3/lib/python3.4/site-packages/twisted/web/http.py", line 1799, in allContentReceived
    req.requestReceived(command, path, version)
  File "/home/hawkowl/venvs/py3/lib/python3.4/site-packages/twisted/web/http.py", line 832, in requestReceived
    self.process()
--- <exception caught here> ---
  File "/home/hawkowl/venvs/py3/lib/python3.4/site-packages/twisted/web/server.py", line 183, in process
    self.render(resrc)
  File "/home/hawkowl/venvs/py3/lib/python3.4/site-packages/twisted/web/server.py", line 234, in render
    body = resrc.render(self)
  File "/home/hawkowl/venvs/py3/lib/python3.4/site-packages/twisted/web/resource.py", line 250, in render
    return m(request)
  File "/home/hawkowl/code/crossbar/crossbar/twisted/resource.py", line 588, in render_GET
    return File.render_GET(self, request)
  File "/home/hawkowl/venvs/py3/lib/python3.4/site-packages/twisted/web/static.py", line 577, in render_GET
    return self.redirect(request)
  File "/home/hawkowl/venvs/py3/lib/python3.4/site-packages/twisted/web/static.py", line 613, in redirect
    return redirectTo(addSlash(request), request)
  File "/home/hawkowl/venvs/py3/lib/python3.4/site-packages/twisted/web/static.py", line 66, in addSlash
    qindex = request.uri.find('?')
builtins.TypeError: 'str' does not support the buffer interface
```

<details><summary>Searchable metadata</summary>

```
trac-id__8169 8169
type__defect defect
reporter__hawkowl hawkowl
priority__high high
milestone__Python_3_x Python-3.x
branch__branches_addSlash_8169_3 branches/addSlash-8169-3
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1452313781517844 1452313781517844
changetime__1455584206390216 1455584206390216
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [46589]) Branching to addSlash-8169.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [46592]) Branching to addSlash-8169-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

I like this, please review.

It also deprecates an unfortunate bit of public API which was busted, and adds a topfile for that too.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Looks good.

Only minor comments.

I don't think that new test does an integration  test for non-ascii paths.
I ran a manual test and looks good.

Maybe instead of

```
base.child('folder').makedirs()
```

we can have

```
base.child(u'fold\u20acr').makedirs()
```

------

Maybe the deprecation news fragment should explicitly state that there is no replacement.

-----

For the new `redirect` method maybe we should add a comment informing that it should be kept in sync with the main Request implementation .

I don't know what are the design requirements for DummyRequest, but maybe we can refactor it as a wrapper around the main Request implementation

```

class DummyRequest(object):

    def __init__(self, *args, **kwargs):
        self._context = Request(*args, **kwargs)

    def getAllHeaders(self):
        return self._context.getAllHeaders()

    def redirect(self, url):
        return self._contenxt.redirect(url)

```

------

Please check my comments and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

This has a duplicate in [#8179](https://github.com/twisted/twisted/issues/8179)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

(In [46799]) Branching to addSlash-8169-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl set status to `closed`|
|-|-|

(In [46802]) Merge addSlash-8169-3: Fix twisted.web.static.addSlash

Author: hawkowl
Reviewer: adiroiban
Fixes: [#8169](https://github.com/twisted/twisted/issues/8169)
